sw/device/silicon_creator/lib/drivers/retention_sram_functest.c

```diff
@@ -0,0 +1,45 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include <stdbool.h>
+#include <stdint.h>
+
+#include "sw/device/lib/runtime/log.h"
+#include "sw/device/lib/runtime/print.h"
+#include "sw/device/silicon_creator/lib/drivers/retention_sram.h"
+#include "sw/device/silicon_creator/lib/error.h"
+#include "sw/device/silicon_creator/lib/test_main.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+
+static rom_error_t retention_sram_clear_test(void) {
+  // Treat the retention SRAM as a raw array of 32-bit words.
+  volatile uint32_t *ret = (volatile uint32_t *)retention_sram_get();
+  const size_t kRetLen = sizeof(retention_sram_t) / sizeof(ret[0]);
+
+  // Set retention SRAM to all non-zero values.
+  for (size_t i = 0; i < kRetLen; ++i) {
+    ret[i] = UINT32_MAX;
+  }
```

Comment:
Isn't this a strict aliasing violation? IIRC you can cast a pointer to a struct to a pointer to the type of its first member but I'm not sure about treating the result as an array (even though we have asserts for the layout of the struct). We could also memset, `retention_sram_clear()`, cast to `char *` and check each byte.

Reply:
I'm not sure. I've converted it to just use the raw retention RAM address and create a `volatile uint32_t*` from that. It's probably a better approach and sidesteps any strict aliasing violations.

I think technically we shouldn't use `memset` because it ignores the `volatile` qualifier.